semantic/src/main/kotlin/tools/samt/semantic/SemanticModel.kt

```diff
@@ -0,0 +1,597 @@
+                                message("Import '${import.name.name}.*' could not be resolved")
+                                highlight("unresolved import", import.location)
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Add all top-level packages to scope
+            val topLevelPackages = global.subPackages.map { PackageType(it) }
+            topLevelPackages.forEach { putIfAbsent(it.packageName, it) }
+
+            // Add built-in types
+            fun addBuiltIn(name: String, type: Type) {
+                putIfAbsent(name, type)?.let { existingType ->
+                    controller.createContext(file.sourceFile).warn {
+                        message("Type '$name' shadows built-in type with same name")
+                        if (existingType is UserDefinedType) {
+                            highlight(existingType.definition.location)
+                        }
+                    }
+                }
+            }
+            addBuiltIn("Int", IntType)
+            addBuiltIn("Long", LongType)
+            addBuiltIn("Float", FloatType)
+            addBuiltIn("Double", DoubleType)
+            addBuiltIn("Decimal", DecimalType)
+            addBuiltIn("Boolean", BooleanType)
+            addBuiltIn("String", StringType)
+            addBuiltIn("Bytes", BytesType)
+            addBuiltIn("Date", DateType)
+            addBuiltIn("DateTime", DateTimeType)
+            addBuiltIn("Duration", DurationType)
+        }
+
+        return FileScope(filePackage, typeLookup)
+    }
+
+    private fun resolveExpression(
+        fileScopes: Map<SourceFile, FileScope>,
+        rootExpression: ExpressionNode,
+    ): ResolvedTypeReference {
+        fun resolveExpression(expression: ExpressionNode): ResolvedTypeReference {
+            val scope = fileScopes.getValue(expression.location.source)
+            when (expression) {
+                is IdentifierNode -> {
+                    scope.typeLookup[expression.name]?.let {
+                        return ResolvedTypeReference(it)
+                    }
+
+                    controller.createContext(expression.location.source).error {
+                        message("Type '${expression.name}' could not be resolved")
+                        highlight("unresolved type", expression.location)
+                    }
+                }
+
+                is BundleIdentifierNode -> {
+                    // Bundle identifiers with one component are treated like normal identifiers
+                    if (expression.components.size == 1) {
+                        return resolveExpression(expression.components.first())
+                    }
+                    // Type is foo.bar.Baz
+                    // Resolve foo first, it must be a package
+                    scope.typeLookup[expression.components.first().name]?.let { expectedPackageType ->
```

Comment:
Couldn't this be done recursively somehow? All the preceding identifier names have to refer to packages anyway